lib/ecto/query/planner.ex

```diff
@@ -307,7 +307,12 @@ defmodule Ecto.Query.Planner do
   defp subquery_source(nil, types), do: {:map, types}
   defp subquery_source(name, types) when is_atom(name), do: {:struct, name, types}
   defp subquery_source({:source, schema, prefix, types}, only) do
-    types = Enum.map(only, fn {field, _} -> {field, Keyword.get(types, field, :any)} end)
+    types =
+      Enum.map(only, fn
+        {field, {:value, type}} -> {field, Keyword.get(types, field, type)}
+        {field, _} -> {field, Keyword.get(types, field, :any)}
```

Comment:
```suggestion
        {field, {:value, type}} -> {field, Keyword.get(types, field, type)}
```
All tests seem pass without this clause. If it is really required maybe it would be good to have a test that needs it

Reply:
Good call! I thought maybe literals wouldn't match this pattern but they do. It seems like the `collect_fields` function is taking care of this.

Reply:
Oh I guess I found one case, when you select a struct:

```elixir
query = from p in Post, join: c in assoc(p, :comments), select_merge: %{text: %Post{}}
```

Now I'm a bit unsure if structs should be disallowed, the same as maps, or if I should add the pattern back.

Reply:
+1 for disallowing structs

Reply:
Done :).  I believe this is a safe way to match the AST for structs, but never did it before so not 100% certain.